klippy/extras/frame_expansion_compensation.py

```diff
@@ -0,0 +1,195 @@
+        self.printer.register_event_handler("homing:homing_move_end",
+                                            self.handle_homing_move_end)
+
+        # Setup temperature sensor
+        self.last_temp = 0.
+        self.measured_min = 99999999.
+        self.measured_max = 0.
+        self.last_home_temp = 0.
+        pheaters = self.printer.load_object(config, 'heaters')
+        self.sensor = pheaters.setup_sensor(config)
+        self.sensor.setup_minmax(self.min_temp, self.max_temp)
+        self.sensor.setup_callback(self.temperature_callback)
+        pheaters.register_sensor(config, self)
+
+        # Z offset transformation
+        self.z_drift_offset = 0.
+        self.comp_enable = True
+        self.last_position = [0., 0., 0., 0.]
+        self.next_transform = None
+        self.comp_state = 'not homed'
+
+        # Register gcode commands
+        self.gcode.register_command('SET_FRAME_COMP',
+                                    self.cmd_SET_FRAME_COMP,
+                                    desc=self.cmd_SET_FRAME_COMP_help)
+        self.gcode.register_command('QUERY_FRAME_COMP',
+                                    self.cmd_QUERY_FRAME_COMP,
+                                    desc=self.cmd_QUERY_FRAME_COMP_help)
+
+    def handle_ready(self):
+        'Called when Klipper successfully connected, ready.'
+        self.toolhead = self.printer.lookup_object('toolhead')
+        gcode_move = self.printer.lookup_object('gcode_move')
+
+        # Z stepper config check
+        kin = self.printer.lookup_object('toolhead').get_kinematics()
+        steppers = [s.get_name() for s in kin.get_steppers()]
+        if not self.z_stepper_name in steppers:
+            raise self.printer.config_error(
+                "'%s' is not a valid stepper" % self.z_stepper_name)
+
+        # Transformation
+        self.next_transform = gcode_move.set_move_transform(self, force=True)
+
+    def get_temp(self, eventtime):
+        'Used by M105, for example, to query probe temperature.'
+        return self.last_temp, 0
+
+    def stats(self, eventtime):
+        return False, '%s: temp=%.1f' % ('Frame', self.last_temp)
+
+    def get_status(self, eventtime):
+        return {
+            'temperature': self.last_temp,
+            'measured_min_temp': self.measured_min,
+            'measured_max_temp': self.measured_max,
+            'current_z_comp': self.z_drift_offset,
+            'frame_ref_temp': self.last_home_temp,
+            'state': self.comp_state
+        }
+
+    def check_eligible(self, endstops):
+        'Check if homed axis is `stepper_z`.'
+        steppers_being_homed = [s.get_name()
+                                for es in endstops
+                                for s in es.get_steppers()]
+        return self.z_stepper_name in steppers_being_homed
+
+    def handle_homing_move_end(self, hmove):
+        'Triggered when Z axis is homed.'
+        if self.check_eligible(hmove.get_mcu_endstops()):
```

Comment:
This should catch the "homing:home_rails_end" event and look for `2 in homing_state.get_axes()`.